### PR TITLE
Replace redux with redux-toolkit createStore

### PR DESCRIPTION
**Pull request by aldovandus**
Pull-Request for `paretOS`

## Description
- Add @reduxjs/toolkit 
- Add configureStore to index.jsx

## Relates to
- #262 Replace redux with redux-toolkit createStore

## Reviewers
- @mikhael28

## Screenshots / other info
...
image
image
Screenshot from 2022-07-02 00-46-43
image

### CODE REVIEW

File: src/index.jsx [hunk 1]
```diff
@@ -32,7 +32,9 @@ if (import.meta.env.NODE_ENV === "production") {
   });
 }
 
-const store = createStore(reducer);
+const store = configureStore({
+  reducer,
+});
```
Comment (mikhael28) [inline src/index.jsx:37]: Hey! Thank you for putting this PR together. A quick note though - as we mentioned in our documentation, we only provide first class support for the `pnpm` package manager. We do not provide support for `yarn` or generic `npm` installs. The reasons for this are explained in the repository readme. Can you please delete the yarn.lock file that you have added here, and please re-install the new dependencies with `pnpm`. You should include changes to the `pnpm-lock` file through this, and not `yarn.lock.` Looking forward to your reply! Please confirm that the app boots up, a screenshot preferably of the code running.

**Review by mikhael28 — CHANGES_REQUESTED** [review]
Please integrate the package with the pnpm package manager.

File: src/index.jsx (hunk 1 above)
Comment (aldovandus) [reply src/index.jsx:37]: Sorry, i didn't read that. I removed yarn.lock and i installed redux toolkit with pnpm.

**Comment by aldovandus** [comment]
I added some screenshots and i installed redux toolkit with pnpm.

**Comment by mikhael28** [comment]
Hey there @aldovandus thank you for doing this! I pulled down the repository to test how the integration of redux-toolkit will affect how we are currently using the Arena, the only place in the repository really where we are using Redux. When testing the actions, it seemed that sometimes it would work (in terms of completing events) but sometimes it wouldn't. I would get this error:
```
immutableStateInvariantMiddleware.ts:25 Uncaught (in promise) Error: Invariant failed: A state mutation was detected inside a dispatch, in the path: sprint.0.teams. Take a look at the reducer(s) handling the action {"type":"COMPLETE_SPRINT_TASK","payload":{"mission":{"_createdAt":"2021-02-01T00:21:04Z","_id":"61d7be0c-3c6b-4c82-9062-9c1dce3636b3","_rev":"iGUy6kFRUd9Q6s0z65x0Pr","_type":"achievement","_updatedAt":"2021-08-17T21:03:12Z","summary":"Accomplish your primary daily objective. Each day, there is something that is truly important. Have you finished that thing?","title":"Eat the Frog","type":"productivity","xp":100,"questions":[],"key":"","img":"","completedAt":1656522160790,"proof":[],"confirmed":false,"completed":true,"description":"Accomplish your primary daily objective. Each day, there is something that is truly important. Have you finished that thing?","proofLink":""},"idx":2,"day":2,"key":"","index":0,"activeSprintIndex":0}}. (https://redux.js.org/style-guide/style-guide#do-not-mutate-state)
    at invariant (immutableStateInvariantMiddleware.ts:25:9)
    at immutableStateInvariantMiddleware.ts:268:11
    at Object.measureTime (utils.ts:10:16)
    at immutableStateInvariantMiddleware.ts:260:20
    at handleChange (Sprint.tsx:113:5)
    at ArenaProofModal.tsx:89:5
    at createFormControl.ts:1048:17
```
I want to pull in @jayeclark here - you have recently-ish done some refactoring work on the Arena, do you see where a state mutation is happening? The local state get's updated, but that error seems to block the web socket transmission of the message. As a result, changes not being saved online/sent to other players.

We only really use Redux in this part of the application, it could be worth having a discussion as to whether we want to replace it with something else. Can you please read https://frontendmastery.com/posts/the-new-wave-of-react-state-management/ and let me know if you have any thoughts?

**Comment by jayeclark** [comment]
Hey @mikhael28, off the top of my head, yeah, there is a direct state mutation in the existing complete sprint task logic (I was just working on that file yesterday converting it to TypeScript) because state is being duplicated via slice rather than via a structured clone. The code block for COMPLETE_SPRINT_TASK  creates a ‘newTeams’ array that copies the existing state - that variable is a new array, but the objects it references are not, so when a team member’s mission is marked as completed, that also changes it’s completion status in state (since both arrays are referencing the same object.)

**Comment by aldovandus** [comment]
Hi, i tried to convert the sprint reducer to the new createReducer  by following this guide:
https://redux-toolkit.js.org/api/createReducer
now it seems to be working well ... I do some tests and can commit the changes.

**Comment by mikhael28** [comment]
@jayeclark plausible to use `lodash.cloneDeep` to get rid of that mutation?

@aldovandus make sure to pull in the latest code from the `main` branch, jaye has done a blinding amount of work in the last day (including converting the reducer to TS)

**Comment by jayeclark** [comment]
@mikhael28 Yeah, I believe that would work.

JS has a native method for deep clones now, but it looks like it doesn’t quite have full browser support yet. https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

**Comment by aldovandus** [comment]
I made a video where I test the application: 

https://user-images.githubusercontent.com/12563425/178059568-e618b583-1d97-4464-aca1-1b8a27ad1dd8.mp4


Let me know if everything is working well.

File: src/arena/Sprint.tsx [hunk 2]
```diff
@@ -93,66 +92,88 @@ function Sprint({ user, history }: SprintProps) {
...
+        payload: {
+          mission,
+          idx,
+          day,
+          key,
+          index: TEAM_INDEX,
+          activeSprintIndex: SPRINT_INDEX,
+        },
+      });
+      const { sprint } = store.getState();
```
[the remark this answers is not in the document]
Comment (mikhael28) [reply src/arena/Sprint.tsx:130]: That makes sense, thank you for explaining - with us being reliant on web socket connection for updates, we are relying at the moment on an 'optimistic' update.

**Review by mikhael28 — APPROVED** [review]
Approved! Sorry for the delay, lots of stuff happened (got COVID, went on vacation, work, etc)

I've resolved a few merge conflicts for you too, to make up for it :)

**Comment by mikhael28** [comment]
@aldovandus would you like to join our contributor-only Discord channel? If so, leave your email here and I will invite you. Devs only.

**Comment by aldovandus** [comment]
Hi,
yes of course. This is my email address: ***@***.***

Il giorno sab 20 ago 2022 alle ore 00:18 mikhael28 ***@***.***>
ha scritto:

**Comment by jeremydthomas** [comment]
@mikhael28

**Comment by mikhael28** [comment]
@aldovandus I'm not sure if you can see it, but your email address is 'blocked out'
Screen Shot 2022-08-25 at 1 42 00 PM
Either repost the email, or send me an email directly to mikhael@hey.com and I will add you then.

**Comment by aldovandus** [comment]
Hi,
i'm sorry. this is the email: aldovandus@gmail.com

**Comment by jeremydthomas** [comment]
I sent you an invite from tommyhilfiger12@hotmail.com @aldovandus

**Pull request merged**